### PR TITLE
[WIP] Use directives in modals

### PR DESCRIPTION
**Pull request by ericjim**
**WIP**

In this PR, we enable users to place directives within modals. Previously, modals were only allowed to take in html strings, thereby limiting their capability to being simple pop-ups. With these changes, it is now possible to add Angular bound logic within the body of a modal (provided that it is a self encapsulated directive/component.

### CODE REVIEW

File: src/layout/index.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ import ModalController from './modal/ws-modal.controller';
 import ModalService from './modal/ws-modal.service';
 import NotificationDirective from './notifications/ws-notification';
 import SecondaryNavDirective from './secondary-nav/ws-secondary-nav';
+import BindHtmlCompile from './modal/bind-html-compile.directive';
 
 const injectorName = 'ws.common.layout';
```
Comment (ericjim) [inline src/layout/index.js]: @SethDavenport 'ws.common.layout' doesn't seem like the right place to be putting this directive. Do you have any suggestions?

Comment (SethDavenport) [reply src/layout/index.js]: Actually I would prefer we take a different approach to this.

Passing HTML as string attributes is a hack I would prefer to not do any more; I have been wanting to deprecate this modal implementation in favour of a declarative one based on [transclusion](https://docs.angularjs.org/api/ng/directive/ngTransclude), which I think would handle your use-case more elegantly.

The old modal is used in a lot of places, let's keep it. But, let's write a new one and do it right now that you have a use case for it.

What do you think about something like this:

```js
angular.module('ws-directives')
  .component('wsModal', {
    bindings: {
      onAction: '&',
      onCancel: '&',
    },
    transclude: {
      trigger: 'wsModalTrigger',
      body: 'wsModalBody',
      action:  '?wsModalAction',
      cancel: '?wsModalCancel',
      progress: '?wsModalProgress',
    }
  });
```

Then you could use it something like this:

```html
<ws-modal
  on-action="$ctrl.somePromiseReturingFunction()"
  on-cancel="$ctrl.someOtherPromiseReturingFunction()">
  <ws-modal-trigger>
    <button>Click me to open a modal</button>
  </ws-modal-trigger>
  <ws-modal-body>
    <p>Whatever HTML you want in here, including other directives</p>
  </ws-modal-body>

   <!-- If omitted, defaults to a simple button with 'go' in it or something-->
  <ws-modal-action>
     <button>Do it!</button>
  </ws-modal-action>

   <!-- If omitted, defaults to a simple button with 'cancel' in it or something-->
  <ws-modal-cancel>
     <button>Actually nope</button>
  </ws-modal-cancel>

   <!-- If omitted, defaults to some text -->
  <ws-modal-progress> 
     <p>Wait for it...</p>
  </ws-modal-progress>
</ws-modal>
```

Comment (SethDavenport) [reply src/layout/index.js]: Take a look at the expander component for an example of multi-slot transclusion.

Comment (ericjim) [reply src/layout/index.js]: 💯  let's build a bigger, badder, better modal.

File: src/layout/modal/bind-html-compile.directive.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+BindHtmlCompile.$inject = ['$compile'];
+
+export default function BindHtmlCompile($compile) {
+  return {
+    restrict: 'A',
+    link: (scope, element, attrs) => {
+      scope.$watch(() => scope.$eval(attrs.bindHtmlCompile),
```
Comment (ericjim) [inline src/layout/modal/bind-html-compile.directive.js]: Dangerous.

**Pull request closed without merging**